src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -507,29 +531,44 @@ public static ClassType GetClassType(Type type, JsonSerializerOptions options)
                 return ClassType.IDictionaryConstructible;
             }
 
```

Comment:
I cleaned up the logic in here so that if something is assignable from IList or IDictionary we always build it as we read. Previously there were many cases where we would flag into IDictionaryConstructible or ICollectionConstructible incorrectly, which is to say we were buffering items into temporary lists where we didn't need to. For scenarios where we now go in-line, that should be a big performance win.